routes/course.js

```diff
@@ -63,57 +67,76 @@ router.get("/api/coursesByTerm/:plan_id", (req, res) => {
     .catch((err) => errorHandler(res, 400, err));
 });
 
-//add course, need to provide course info as json object in request body
-//distribution field is also updated
+// add course, need to provide course info as json object in request body
+// year, distribution, and fineReq objects updated
 router.post("/api/courses", async (req, res) => {
-  const course = req.body;
-  //console.log("course is ", course);
-  await plans
-    .findById(course.plan_id)
-    .then((plan) => {
-      //console.log(plan);
-      course.distribution_ids.forEach((id) => {
-        if (!plan.distribution_ids.includes(id)) {
-          errorHandler(res, 400, {
-            message: "Invalid combination of plan_id and distribution_ids.",
+  const courseBody = req.body;
+  await courses
+    .create(courseBody)
+    .then(async (retrievedCourse) => {
+      // find year obj and insert course id to array
+      await years
+        .findOneAndUpdate(
+          {
+            $and: [
+              { plan_id: retrievedCourse.plan_id },
+              { name: retrievedCourse.year },
+            ],
+          },
+          { $push: { courses: retrievedCourse._id } },
+          { new: true, runValidators: true }
+        )
+        .then(async (year) => {
+          retrievedCourse.year_id = year._id; // set year_id
+          await retrievedCourse.save();
+        });
+      // update plan's distribution objs
+      const plan = await plans.findById(retrievedCourse.plan_id);
+      for (let m_id of plan.major_ids) {
+        let distDoubleCount = undefined;
+        await distributions
+          .find({ plan_id: retrievedCourse.plan_id, major_id: m_id })
+          .then(async (distObjs) => {
+            for (let distObj of distObjs) {
+              if (
+                !distObjs.satisfied &&
```

Comment:
consider the scenario where a distribution is satisfied but the fine requirements are not. In this case the new course would never be able to go into this distribution and be checked for its fine requirement.

Reply:
we should keep checking for the distribution even they have been satisfied.

Reply:
That is true. Then we also have to make sure the planned credits field is correct because if `planned` credits stop updating once it meets `required_credits`.

I'm not sure whether to let `planned` overflow, or just recalculate linearly whenever we add / delete a course

Reply:
But yes ultimately, we need to keep checking for distributions, even if satisfied!!